packages/mask/src/plugins/Referral/Worker/apis/indexers.ts

```diff
@@ -0,0 +1,154 @@
+
+        if (nearest) {
+            return nearest.n
+        }
+    }
+
+    // No nearest found, it's up to the caller to implement another node selection strategy.
+    return null
+}
+
+export function getNearestAndRemainingNodes(
+    nodes: Node[],
+    airports: Airport[],
+    pop: string,
+    nearestNodeAmount: number,
+) {
+    let remainingNodes = nodes
+
+    // get N nearest nodes
+    const nearestNodes: Node[] = []
+    for (let i = 0; i < nearestNodeAmount; i += 1) {
+        const nearestNode = selectNearestNode(remainingNodes, airports, pop)
+
+        if (nearestNode) {
+            nearestNodes.push(nearestNode)
+            // filter out a node that is in nearest nodes already
+            remainingNodes = remainingNodes.filter((node) => node.url !== nearestNode.url)
+        }
+    }
+
+    return {
+        nearestNodes,
+        remainingNodes,
+    }
+}
+
+async function findQuorum(nodes: Node[], airports: Airport[], pop: string, urlPath: string, minQuorum: number) {
+    let responses: any[] = []
+    let nodeCount = minQuorum
+
+    // query another nodes among the remaining nodes until quorum is reached
+    while (responses.length < minQuorum) {
+        if (!nodes.length) {
+            break
+        }
+
+        const { remainingNodes, nearestNodes } = getNearestAndRemainingNodes(nodes, airports, pop, nodeCount)
+
+        const settledResponses = await Promise.allSettled(
+            nearestNodes.map(async (node) => await (await fetch(`${node.url}${urlPath}`)).json()),
+        )
+        // filter out failed  responses
+        responses = [...responses, ...settledResponses.filter((response) => response.status === 'fulfilled')]
+
+        nodeCount = minQuorum - responses.length
+        nodes = remainingNodes
+    }
+
+    return responses
+}
+
+export async function queryIndexersWithNearestQuorum(searchParams: LogParams): Promise<any> {
+    const urlPath = makeIndexerUrlPath(searchParams)
+    const { discovery, pop } = await getDiscovery()
+    const minQuorum = 1
+
+    const responses = await findQuorum(discovery.indexers, discovery.airports, pop, urlPath, minQuorum)
+
+    if (!responses.length) {
+        throw new Error('Node request error')
+    }
+
+    return responses[0].value
+}
+
+interface LogParams {
+    // Logs from these addresses only
+    addresses?: EvmAddress[]
+    // Search across all topics
+    topics?: Bytes32[]
+    // Topic1 is commonly the event id
+    topic1?: Bytes32[]
+    // Topics 2-4 are the 1-3 indexed event params
+    topic2?: Bytes32[]
+    topic3?: Bytes32[]
+    topic4?: Bytes32[]
+    // Logs from these chains only
+    chainId?: number[]
+    // Logs from these transactions only
+    transactionHash?: Bytes32[]
+    // Default AND search is done by the indexers
+    strategy?: 'AND' | 'OR'
+}
+
+export function makeIndexerUrlPath(params: LogParams) {
+    const parts = []
+    if (params.addresses) params.addresses.forEach((d) => parts.push(`address=${d}`))
```

Comment:
https://www.npmjs.com/package/urlcat

Reply:
[urlcat](https://www.npmjs.com/package/urlcat) doesn't work for this case, because params have array values:
For example, topic2= ["0xf9512da9b772c1af62345b0192609e99d11dd28ff3f3634f3759c5280c20e290", "0xc5d7ef5d03798b55b1a2471559ec455c0dcebbdb3c11e27ed4dac94de9d22ece"]

urlcat returns:
`baseTemplate/path?topic2=0xf9512da9b772c1af62345b0192609e99d11dd28ff3f3634f3759c5280c20e290%2C0xc5d7ef5d03798b55b1a2471559ec455c0dcebbdb3c11e27ed4dac94de9d22ece`

API requires:
`baseTemplate/path?topic2=0xf9512da9b772c1af62345b0192609e99d11dd28ff3f3634f3759c5280c20e290&topic2=C0xc5d7ef5d03798b55b1a2471559ec455c0dcebbdb3c11e27ed4dac94de9d22ece`